Add configuration panel for Application Credentials

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
Add a configuration panel for Application Credentials which allow users to configure OAuth Client ID and OAuth Client
Secret for an integration using the user interface.

Work in progress:
- [x] determine correct place for this in the UI (arbitrary placeholder for now, asked for input in `#devs_us`)
- [x] add core support for an integration (e.g. xbox)
- [x] add core websocket for returning supported integrations
- [x] render integrations drop down or search in fronted UI
- [x] support for deleting an integration
- [ ] special display of imported credentials vs user added credentials

<img width="650" alt="Capture" src="https://user-images.githubusercontent.com/6026418/163735682-95ca536d-f7f4-404f-9aa9-dea590b67d9a.PNG">


## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion: https://github.com/home-assistant/core/pull/69148 https://github.com/home-assistant/architecture/discussions/692
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

zsarnett commented:
@matthiasdebaat Let's talk about this. Not sure where this should be. The configuration sections are all changing so this will definitely not stay where is but maybe move to the system tab 🤷

allenporter commented:
> @matthiasdebaat Let's talk about this. Not sure where this should be. The configuration sections are all changing so this will definitely not stay where is but maybe move to the system tab 🤷

Thanks, really appreciate this.  I'll continue to work on the functionality, thought just wanted to flag this early.

allenporter commented:
> > @matthiasdebaat Let's talk about this. Not sure where this should be. The configuration sections are all changing so this will definitely not stay where is but maybe move to the system tab 🤷
> 
> Thanks, really appreciate this. I'll continue to work on the functionality, thought just wanted to flag this early.

@zsarnett i saw you landed a config revamp. Do you have a more well formed opinion now of where this should land, or otherwise have any principles regarding the most recent revamp?

I still have other pieces to iterate on fix up but thought I would ask while this was top of mind.

allenporter commented:
> > > @matthiasdebaat Let's talk about this. Not sure where this should be. The configuration sections are all changing so this will definitely not stay where is but maybe move to the system tab 🤷
> > 
> > 
> > Thanks, really appreciate this. I'll continue to work on the functionality, thought just wanted to flag this early.
> 
> @zsarnett i saw you landed a config revamp. Do you have a more well formed opinion now of where this should land, or otherwise have any principles regarding the most recent revamp?
> 
> I still have other pieces to iterate on fix up but thought I would ask while this was top of mind.

I just rebased and can see how `System` makes sense given it seems like we've found a place to put low level things like this. I'm going to go ahead and move this there optimistically, given I think I can see where you're going with this, but happy to put this wherever makes sense.

zsarnett commented:
Similar to the help icon except you will need to use an overflow button

https://github.com/home-assistant/frontend/blob/dev/src/panels/config/automation/ha-automation-picker.ts#L246-L251

Overflow Icon Button
https://github.com/home-assistant/frontend/blob/dev/src/panels/config/dashboard/ha-config-dashboard.ts#L157-L171

zsarnett commented:
```html
  <ha-button-menu
    corner="BOTTOM_START"
    slot="toolbar-icon"
    activatable
  >
    <ha-icon-button
      slot="trigger"
      .label=${this.hass.localize("ui.common.menu")}
      .path=${mdiDotsVertical}
    ></ha-icon-button>
    <ha-clickable-list-item
      @click=${this._entryClicked}
      href="/config/lovelace/resources"
      aria-label=${this.hass.localize(
        "ui.panel.config.lovelace.resources.caption"
      )}
    >
      ${this.hass.localize(
        "ui.panel.config.lovelace.resources.caption"
      )}
    </ha-clickable-list-item>
  </ha-button-menu>
```

allenporter commented:
Thank you that was extremely helpful. I am hitting a couple layout issues, still working through.

The overflow menu looks good ✅ 
<img width="868" alt="Overflow menu" src="https://user-images.githubusercontent.com/6026418/164918145-5dde4740-6e5f-4544-bd64-4527b444479f.png">

though it does not expand quite right ❓ 
<img width="1440" alt="Overflow menu expanded issue" src="https://user-images.githubusercontent.com/6026418/164918308-bbf2674d-2562-4c43-b8b0-625474eaba41.png">

And in narrow mode it looks good on 3 tabs ✅ 
<img width="1440" alt="Narrow overflow" src="https://user-images.githubusercontent.com/6026418/164918875-1374c70a-dd32-4433-b948-95abfa5fb53d.png">

However on the integrations tab it gets out of place ❌ 
<img width="667" alt="Narrow overflow integrations issue" src="https://user-images.githubusercontent.com/6026418/164919128-12792cbe-aa72-41de-989a-5e90d341f0b2.png">

zsarnett commented:
I'll try to take a look this weekend or Monday

allenporter commented:
No rush, I also need to do one last selector for an integration drop down

matthiasdebaat commented:
Awesome work @allenporter!

zsarnett committed:
get back to allens commits

zsarnett committed:
Update Overflow menu

zsarnett commented:
I messed up the git commands but should be all fixed up now. Your commits are all the first commit now... Sorry about that. But the overflow should be good

allenporter commented:
@zsarnett 🔥 🔥🔥🔥!!!! thank you, so kind of you. I now will work on making the integration selector and take out of draft.

allenporter committed:
Load integration selector and pre-load translations

allenporter committed:
Address lint errors from `yarn run lint:types`

zsarnett commented:
When the backend gets merged can you set it back to ready for review?

zsarnett commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,233 @@
+  name: string;
+}
+
+const rowRenderer: ComboBoxLitRenderer<Domain> = (item) => html`<mwc-list-item>
```
`mwc-list-item` Needs imported

allenporter replied:
Imported.

zsarnett commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,233 @@
+        this._createApplicationCredential(ev);
+      }
+    });
+    this._fetchConfig();
```
can we do this in `showDialog()`

allenporter replied:
Done.

zsarnett commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,233 @@
+            required
+            @value-changed=${this._handleDomainPicked}
+          ></ha-combo-box>
+          <paper-input
```
Use `ha-textfield`

allenporter replied:
Fixed.

zsarnett commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,233 @@
+            dialogInitialFocus
+          ></paper-input>
+
+          <paper-input
```
Use `ha-textfield`

allenporter replied:
Fixed.

zsarnett commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,233 @@
+    `;
+  }
+
+  private _close() {
```
Set domains to undefined

allenporter replied:
Fixed.

zsarnett commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,233 @@
+import "../../../components/ha-combo-box";
+import { ComboBoxLitRenderer } from "lit-vaadin-helpers";
+import { createCloseHeading } from "../../../components/ha-dialog";
+import "../../../components/ha-formfield";
+import "../../../components/ha-switch";
```
Unused

allenporter replied:
Removed.

zsarnett commented on src/panels/config/application_credentials/ha-config-application-credentials.ts:
```diff
@@ -0,0 +1,258 @@
+                          .path=${mdiDelete}
+                          .label=${this.hass.localize("ui.common.remove")}
+                        ></ha-icon-button>
+                        <paper-tooltip animation-delay="0" for="remove-btn">
```
use `ha-help-tooltip` or import `paper-tooltip`

allenporter replied:
Done

zsarnett commented on src/panels/config/entities/entity-registry-settings.ts:
```diff
@@ -276,7 +276,6 @@ export class EntityRegistrySettings extends SubscribeMixin(LitElement) {
                 @selected=${this._deviceClassChanged}
                 @closed=${stopPropagation}
               >
-                <mwc-list-item></mwc-list-item>
```
hmm this got carried over it seems. Can you add this back.

allenporter replied:
Whoops!

zsarnett commented on src/panels/config/integrations/ha-integration-overflow-menu.ts:
```diff
@@ -0,0 +1,49 @@
+  private _entryClicked(ev) {
+    ev.currentTarget.blur();
+  }
+
+  static get styles() {
+    return css``;
+  }
```
```suggestion
```

allenporter committed:
Address pr feedback

allenporter committed:
Address pr feedback

allenporter committed:
Update src/panels/config/integrations/ha-integration-overflow-menu.ts

Co-authored-by: Zack Barett <zackbarett@hey.com>

allenporter committed:
Move fetchconfig to show dialog

allenporter committed:
Switch to ha-help-tooltip

allenporter committed:
Remove unused import

allenporter commented:
> When the backend gets merged can you set it back to ready for review?

Core PR said: `Great! Can be merged when frontend PR is happy with this.`
https://github.com/home-assistant/core/pull/69148#pullrequestreview-955848132

So figured i'd get your pulse before actually merging.

bramkragten commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,229 @@
+    `;
+  }
+
+  private _close() {
```
This should be a public function and fire `dialog-closed` event:

```ts
  public closeDialog() {
    this._params = undefined;
    this._domains = undefined;
    fireEvent(this, "dialog-closed", { dialog: this.localName });
  }
```

allenporter replied:
Fixed.

allenporter committed:
Rename close to public closeDialog function

allenporter committed:
Fix erros added on move to ha-textfield

bramkragten commented on src/panels/config/application_credentials/dialog-add-application-credential.ts:
```diff
@@ -0,0 +1,240 @@
+
+  protected firstUpdated(changedProperties: PropertyValues) {
+    super.firstUpdated(changedProperties);
+    this.addEventListener("keypress", (ev) => {
```
I think this is handled by the dialog, and this is not needed?

allenporter replied:
Removed.

allenporter committed:
Remove unnecessary dialog keystoke handling

allenporter committed:
Run prettier

yarn prettier --write src/panels/config/application_credentials/dialog-add-application-credential.ts

allenporter commented:
Whoops, noticed I had a lint error, fixed.
